src/lib/provable/crypto/foreign-twisted.ts

```diff
@@ -0,0 +1,309 @@
+  }
+
+  /**
+   * Twisted elliptic curve scalar multiplication, where the scalar is represented as a {@link ForeignField} element.
+   *
+   * **Important**: this proves that the result of the scalar multiplication is not the zero point.
+   *
+   * @throws if the scalar multiplication results in the zero point; for example, if the scalar is zero.
+   *
+   * @example
+   * ```ts
+   * let r = p.scale(s); // r = s * p
+   * ```
+   */
+  scale(scalar: AlmostForeignField | bigint | number) {
+    let Curve = this.Constructor.Bigint;
+    let scalar_ = this.Constructor.Scalar.from(scalar);
+    let p = CurveTwisted.scale(scalar_.value, toPoint(this), Curve);
+    return new this.Constructor(p);
+  }
+
+  static assertOnCurve(g: ForeignTwisted) {
+    CurveTwisted.assertOnCurve(toPoint(g), this.Bigint);
+  }
+
+  /**
+   * Assert that this point lies on the elliptic curve, which means it satisfies the equation
+   * `a * x^2 + y^2 = 1 + d * x^2 * y^2`
+   */
+  assertOnCurve() {
+    this.Constructor.assertOnCurve(this);
+  }
+
+  // dynamic subclassing infra
+  get Constructor() {
+    return this.constructor as typeof ForeignTwisted;
+  }
+  static _Bigint?: CurveTwisted;
+  static _Field?: typeof AlmostForeignField;
+  static _Scalar?: typeof AlmostForeignField;
+  static _provable?: ProvablePureExtended<
+    ForeignTwisted,
+    { x: bigint; y: bigint },
+    { x: string; y: string }
+  >;
+
+  /**
+   * Curve arithmetic on JS bigints.
+   */
+  static get Bigint() {
+    assert(this._Bigint !== undefined, 'ForeignTwisted not initialized');
+    return this._Bigint;
+  }
+  /**
+   * The base field of this curve as a {@link ForeignField}.
+   */
+  static get Field() {
+    assert(this._Field !== undefined, 'ForeignTwisted not initialized');
+    return this._Field;
+  }
+  /**
+   * The scalar field of this curve as a {@link ForeignField}.
+   */
+  static get Scalar() {
+    assert(this._Scalar !== undefined, 'ForeignTwisted not initialized');
+    return this._Scalar;
+  }
+  /**
+   * `Provable<ForeignCurve>`
+   */
+  static get provable() {
+    assert(this._provable !== undefined, 'ForeignTwistednot initialized');
+    return this._provable;
+  }
+}
+
+/**
+ * Create a class representing a twisted elliptic curve group, which is different from the native {@link Group}.
+ *
+ * ```ts
+ * const Curve = createForeignTwisted(Crypto.TwistedCurveParams.Edwards25519);
+ * ```
+ *
+ * `createForeignTwisted(params)` takes curve parameters {@link TwistedCurveParams} as input.
+ * We support `modulus` and `order` to be prime numbers up to 259 bits.
+ *
+ * The returned {@link ForeignCurveNotNeeded} class represents a _non-zero curve point_ and supports standard
+ * elliptic curve operations like point addition and scalar multiplication.
+ *
+ * {@link ForeignCurveNotNeeded} also includes to associated foreign fields: `ForeignCurve.Field` and `ForeignCurve.Scalar`, see {@link createForeignField}.
```

Comment:
"ForeignCurveNotNeeded"? :D

> represents a _non-zero curve point_

why only a non-zero point? I think all points can be represented?

Reply:
> "ForeignCurveNotNeeded"? :D

https://github.com/o1-labs/o1js/commit/10cfa8a8a59625b37515722d53630464b0ff4da3#diff-e2b15bfc1f6103679925c6ccf6697f5877d73e219b0433976dfba6111cf45febR306

👀 😁

Reply:
haha damn